rpc/src/p2p.rs

```diff
@@ -1,17 +1,207 @@
+    #[serde(rename = "NumStreamsInbound")]
+    pub streams_inbound: u32,
+    #[serde(rename = "NumStreamsOutbound")]
+    pub streams_outbound: u32,
+    #[serde(rename = "NumConnsInbound")]
+    pub connections_inbound: u32,
+    #[serde(rename = "NumConnsOutbound")]
+    pub connections_outbound: u32,
+    #[serde(rename = "NumFD")]
+    pub file_descriptors: u32,
+    #[serde(rename = "Memory")]
+    pub memory: u32,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct ResourceManagerStats {
+    #[serde(rename = "System")]
+    pub system: Stat,
+    #[serde(rename = "Transient")]
+    pub transient: Stat,
+    #[serde(rename = "Services")]
+    pub services: HashMap<String, Stat>,
+    #[serde(rename = "Protocols")]
+    pub protocols: HashMap<String, Stat>,
+    #[serde(rename = "Peers")]
+    pub peers: HashMap<String, Stat>,
+}
+
+#[derive(Debug, Error)]
+pub enum PeerIdError {
+    #[error("unable to decode base58 string")]
+    Base58Error,
+    #[error("libp2p error")]
+    Libp2pError(#[from] libp2p_identity::ParseError),
+}
+
+#[derive(Debug, PartialEq, Eq)]
+pub struct PeerId(pub libp2p::PeerId);
+
+impl<'de> Deserialize<'de> for PeerId {
+    fn deserialize<D>(deserializer: D) -> Result<PeerId, D::Error>
+    where
+        D: de::Deserializer<'de>,
+    {
+        let s: String = Deserialize::deserialize(deserializer)?;
+        s.parse().map_err(D::Error::custom)
+    }
+}
+
+impl Serialize for PeerId {
+    fn serialize<S>(&self, s: S) -> Result<S::Ok, S::Error>
+    where
+        S: ser::Serializer,
+    {
+        s.serialize_str(&self.0.to_base58())
+    }
+}
+
+impl FromStr for PeerId {
+    type Err = PeerIdError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let bytes = bs58::decode(s)
+            .into_vec()
+            .map_err(|_| PeerIdError::Base58Error)?;
+
+        Ok(Self(libp2p::PeerId::from_bytes(&bytes)?))
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct BandwidthStats {
+    #[serde(rename = "TotalIn")]
+    pub total_in: f32,
+
+    #[serde(rename = "TotalOut")]
+    pub total_out: f32,
+
+    #[serde(rename = "RateIn")]
+    pub rate_in: f32,
+
+    #[serde(rename = "RateOut")]
+    pub rate_out: f32,
+}
+
+#[derive(Debug, PartialEq, Eq)]
+pub enum Connectedness {
+    NotConnected = 0,
+    Connected = 1,
+    CanConnect = 2,
+    CannotConnect = 3,
+}
+
+impl<'de> Deserialize<'de> for Connectedness {
+    fn deserialize<D>(deserializer: D) -> Result<Connectedness, D::Error>
+    where
+        D: de::Deserializer<'de>,
+    {
+        let c: u32 = Deserialize::deserialize(deserializer)?;
+        match c {
+            0 => Ok(Connectedness::NotConnected),
+            1 => Ok(Connectedness::Connected),
+            2 => Ok(Connectedness::CanConnect),
+            3 => Ok(Connectedness::CannotConnect),
+            v => Err(D::Error::unknown_variant(
+                &v.to_string(),
+                &["0", "1", "2", "3"],
+            )),
+        }
+    }
+}
```

Comment:
I think this can be derived. check other places too

Reply:
I added manual implementation because default `serde` one assumes that for enums, json representation is a string/map. Thankfully active community member already created a [`crate`](https://github.com/dtolnay/serde-repr) just for this situation. Hopefully there are no sneaky binary blobs inside :)

Added automatic derivation using [`serde_repr`](https://github.com/dtolnay/serde-repr) crate.